community/kernel/src/main/java/org/neo4j/kernel/api/DataWriteOperations.java

```diff
@@ -21,16 +21,17 @@
 
 import java.util.Map;
 
-import org.neo4j.kernel.api.exceptions.EntityNotFoundException;
-import org.neo4j.kernel.api.exceptions.InvalidTransactionTypeKernelException;
+import org.neo4j.internal.kernel.api.ExplicitIndexWrite;
+import org.neo4j.internal.kernel.api.exceptions.InvalidTransactionTypeKernelException;
 import org.neo4j.internal.kernel.api.exceptions.KernelException;
+import org.neo4j.internal.kernel.api.exceptions.explicitindex.AutoIndexingKernelException;
+import org.neo4j.internal.kernel.api.exceptions.explicitindex.ExplicitIndexNotFoundKernelException;
+import org.neo4j.kernel.api.exceptions.EntityNotFoundException;
 import org.neo4j.kernel.api.exceptions.RelationshipTypeIdNotFoundKernelException;
-import org.neo4j.kernel.api.exceptions.explicitindex.AutoIndexingKernelException;
-import org.neo4j.kernel.api.exceptions.explicitindex.ExplicitIndexNotFoundKernelException;
 import org.neo4j.kernel.api.exceptions.schema.ConstraintValidationException;
 import org.neo4j.values.storable.Value;
 
-public interface DataWriteOperations
+public interface DataWriteOperations extends ExplicitIndexWrite
```

Comment:
I'm not completely up to date with the problem, but if we have direct dependencies on kernel exceptions in a legacy cypher version that we support, we must patch that neo4j to remove the dependency before we can move it in 3.4. We will act on this eventually.

Reply:
The plan is for DataWriteOperations to die once we are done porting I add the `extends ExplicitIndexWrite` so that we don't have to have two of every method in `AutoIndexOperations` - on for `ExplicitIndexWrite` and one for `DataWriteOperations`.  We cannot remove it until we are done porting relationship stuff as well, but I promise we'll get there eventually.

Yes you are right they are not identical since we have an unfortunate dependency from cypher that makes it hard to move `EntitiyNotFoundException`, in order to move it we need to first fix it in 2.3, 3.1, and 3.2 and then wait for them to be released so we can update the binary dependency.

Reply:
Sound like you are on top of this! I'm happy 🎉